File: app/src/main/java/closer/vlllage/com/closer/handler/bubble/MapBubbleEventView.java
package closer.vlllage.com.closer.handler.bubble;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.handler.event.EventDetailsHandler;
import closer.vlllage.com.closer.handler.helpers.OutboundHandler;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.models.Event;

public class MapBubbleEventView extends PoolMember {

    public View from(ViewGroup layer, MapBubble mapBubble, MapBubbleEventClickListener onClickListener) {
        View view = LayoutInflater.from(layer.getContext()).inflate(R.layout.map_bubble_event, layer, false);

        view.findViewById(R.id.click).setOnClickListener(v -> onClickListener.onEventClick(mapBubble));
        update(view, mapBubble);

        view.findViewById(R.id.directionsButton).setOnClickListener(v -> $(OutboundHandler.class).openDirections(mapBubble.getLatLng()));

        return view;
    }

    public void update(View view, MapBubble mapBubble) {
        TextView bubbleTextView = view.findViewById(R.id.bubbleText);
        TextView actionTextView = view.findViewById(R.id.action);
        bubbleTextView.setText(mapBubble.getStatus());

        actionTextView.setText($(EventDetailsHandler.class).formatEventDetails((Event) mapBubble.getTag()));
    }

    public interface MapBubbleEventClickListener {
        void onEventClick(MapBubble mapBubble);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/bubble/MapBubbleEventView.kt
package closer.vlllage.com.closer.handler.bubble

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.handler.event.EventDetailsHandler
import closer.vlllage.com.closer.handler.helpers.OutboundHandler
import com.queatz.on.On
import closer.vlllage.com.closer.store.models.Event

class MapBubbleEventView constructor(private val on: On) {

    fun from(layer: ViewGroup, mapBubble: MapBubble, onClickListener: MapBubbleEventClickListener): View {
        val view = LayoutInflater.from(layer.context).inflate(R.layout.map_bubble_event, layer, false)

        view.findViewById<View>(R.id.click).setOnClickListener { onClickListener.invoke(mapBubble) }
        update(view, mapBubble)

        view.findViewById<View>(R.id.directionsButton).setOnClickListener { on<OutboundHandler>().openDirections(mapBubble.latLng) }

        return view
    }

    fun update(view: View, mapBubble: MapBubble) {
        val bubbleTextView = view.findViewById<TextView>(R.id.bubbleText)
        val actionTextView = view.findViewById<TextView>(R.id.action)
        bubbleTextView.text = mapBubble.status

        if ((mapBubble.tag as Event).isPublic) {
            bubbleTextView.setCompoundDrawablesRelativeWithIntrinsicBounds(0, 0, 0, 0)
        } else {
            bubbleTextView.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.ic_group_black_18dp, 0, 0, 0)
        }

        actionTextView.text = on<EventDetailsHandler>().formatEventDetails(mapBubble.tag as Event)
    }
}

typealias MapBubbleEventClickListener = (mapBubble: MapBubble) -> Unit
